### PR TITLE
ReverseProtoWriter: like ProtoWriter but without measurement

### PR DESCRIPTION
**Pull request by swankjesse**
In performance measurement I observed Wire's encodedSize() showing up too much in
profiles. This is likely due to the fact that this method must be invoked repeatedly
for deeply-nested datastructures.

Attempt to improve things by changing encoding from front-to-back to back-to-front,
which allows us to capture the data size for free as a side-effect.

### CODE REVIEW

File: wire-library/wire-java-generator/src/main/java/com/squareup/wire/java/JavaGenerator.java [hunk 1]
```diff
@@ -886,17 +887,19 @@ private TypeSpec enumAdapter(NameAllocator nameAllocator, EnumType type, ClassNa
...
-        .addModifiers(PUBLIC)
-        .addParameter(ProtoWriter.class, writer)
-        .addParameter(javaType, value)
-        .addException(IOException.class)
-        .addStatement("int $N = toValue($N)", i, value)
-        .addStatement("if ($N == $L) throw new $T($S + $N)",
-            i, -1, ProtocolException.class, "Unexpected enum constant: ", value)
-        .addStatement("$N.writeVarint32($N)", writer, i)
-        .build());
+    for (boolean reverse : Arrays.asList(false, true)) {
```
Comment (oldergod) [inline wire-library/wire-java-generator/src/main/java/com/squareup/wire/java/JavaGenerator.java:890]: maybe we could locally define a method and call it twice in a row with a named parameter to make it more reasonable

Comment (oldergod) [reply wire-library/wire-java-generator/src/main/java/com/squareup/wire/java/JavaGenerator.java:890]: hmmmm Java™

**Review by oldergod — APPROVED** [review]
great tests

File: wire-library/wire-java-generator/src/main/java/com/squareup/wire/java/JavaGenerator.java (hunk 1 above)
Comment (swankjesse) [reply wire-library/wire-java-generator/src/main/java/com/squareup/wire/java/JavaGenerator.java]: Fixed!

**Pull request merged**